### ISSUE TITLE
internal error: at cranko-0.8.0\src\main.rs:374:56

### ISSUE THREAD

**Issue by oovm** [pkgw/cranko#24]
## Problem

run `cranko confirm --force` at my project [panduck](https://github.com/notedge/panduck) entered unreachable code

I haven't pushed yet, this is the local version where the problem occurred [panduck.zip](https://github.com/notedge/panduck-cli/releases/download/v0/panduck.zip)

## Full log

```rs
Administrator@Aster-Surface MINGW64 /d/Rust/panduck (main)
$ RUST_BACKTRACE=full cranko confirm --force
warning: cannot find Cargo version requirement for dependency of `panduck-ebook` on `panduck_core`
warning: missing or invalid key `internal_dep_versions.panduck_core` in `D:\Rust\panduck\projects\panduck-ebook\Cargo.toml`
warning: ... this is needed to specify the oldest version of `panduck_core` compatible with `panduck-ebook`
warning: cannot find Cargo version requirement for dependency of `panduck-html` on `panduck_core`
warning: missing or invalid key `internal_dep_versions.panduck_core` in `D:\Rust\panduck\projects\panduck-html\Cargo.toml`
warning: ... this is needed to specify the oldest version of `panduck_core` compatible with `panduck-html`
warning: cannot find Cargo version requirement for dependency of `panduck-pdf` on `panduck_core`
warning: missing or invalid key `internal_dep_versions.panduck_core` in `D:\Rust\panduck\projects\panduck-pdf\Cargo.toml`
warning: ... this is needed to specify the oldest version of `panduck_core` compatible with `panduck-pdf`
warning: cannot find Cargo version requirement for dependency of `panduck-sycamore` on `panduck_core`
warning: missing or invalid key `internal_dep_versions.panduck_core` in `D:\Rust\panduck\projects\panduck-sycamore\Cargo.toml`
warning: ... this is needed to specify the oldest version of `panduck_core` compatible with `panduck-sycamore`
warning: cannot find Cargo version requirement for dependency of `panduck-sycamore` on `panduck_html`
warning: missing or invalid key `internal_dep_versions.panduck_html` in `D:\Rust\panduck\projects\panduck-sycamore\Cargo.toml`
warning: ... this is needed to specify the oldest version of `panduck_html` compatible with `panduck-sycamore`
warning: cannot find Cargo version requirement for dependency of `panduck-web` on `panduck_core`
warning: missing or invalid key `internal_dep_versions.panduck_core` in `D:\Rust\panduck\projects\panduck-web\Cargo.toml`
warning: ... this is needed to specify the oldest version of `panduck_core` compatible with `panduck-web`
warning: not recommended to confirm with a modified working tree (the file backing repository is dirty: file projects/panduck-core/Cargo.toml has been modified)
info: panduck-zola: micro bump (expected: 0.13.0 => 0.13.1)
info: panduck-yew: micro bump (expected: 0.18.0 => 0.18.1)
info: panduck-core: micro bump (expected: 0.1.3 => 0.1.4)
info: panduck-web: micro bump (expected: 0.4.0 => 0.4.1)
thread 'main' panicked at 'internal error: entered unreachable code', C:\Users\Administrator\.cargo\registry\src\github.com-1ecc6299db9ec823\cranko-0.8.0\src\main.rs:374:56
stack backtrace:
   0:     0x7ff74a75360f - git_patch_owner
   1:     0x7ff74a772fea - git_patch_owner
   2:     0x7ff74a74aa08 - git_patch_owner
   3:     0x7ff74a755d96 - git_patch_owner
   4:     0x7ff74a75587c - git_patch_owner
   5:     0x7ff74a7563f5 - git_patch_owner
   6:     0x7ff74a755faf - git_patch_owner
   7:     0x7ff74a753f37 - git_patch_owner
   8:     0x7ff74a755f39 - git_patch_owner
   9:     0x7ff74a8742a0 - git_proxy_options_init
  10:     0x7ff74a8741ec - git_proxy_options_init
  11:     0x7ff74a376dff - <unknown>
  12:     0x7ff74a3916cb - <unknown>
  13:     0x7ff74a392d11 - <unknown>
  14:     0x7ff74a2fa786 - <unknown>
  15:     0x7ff74a32849c - <unknown>
  16:     0x7ff74a752f57 - git_patch_owner
  17:     0x7ff74a39e047 - <unknown>
  18:     0x7ff74a85f6f4 - git_proxy_options_init
  19:     0x7ffcf2137034 - BaseThreadInitThunk
  20:     0x7ffcf3062651 - RtlUserThreadStart
```

**Comment by pkgw** [comment]
Thanks for trying Cranko and for reporting this issue! My apologies for being slow in getting back to you.

It looks like the error you're encountering is related to the missing version requirements in the internal dependencies in your project, as per the warning messages earlier in the output. Of course, the program is supposed to exit with an error rather than just crashing as you've encountered. Maybe the `--force` option is causing it to not give up as early as it should in your particular situation?

To be honest my bandwidth to work on Cranko is very limited so I don't expect to be able to spend much time on fixing this, but I'll certainly review a pull request if you can put one together. And I strongly suspect that if you specify the internal version requirements in your `Cargo.toml`, the crash will go away. Unfortunately, looking over the book, I don't seem to have documented this aspect of the workflow very well at all :-( Here's some discussion of internal dependency versioning for *Python* projects: [link](https://pkgw.github.io/cranko/book/latest/integrations/python.html#versioning-internal-dependencies). And here's an example of how it's done in the wild with Rust: [link](https://github.com/tectonic-typesetting/tectonic/blob/master/Cargo.toml#L129). Finally, the last two paragraphs of [this book section](https://pkgw.github.io/cranko/book/latest/jit-versioning/index.html#the-monorepo-wrinkle) give some of the high-level motivation for the approach.

**Comment by pkgw** [comment]
Well, as it happens, I ran into this problem today as I was working on the first new Cranko feature in a while. I think I just had an oversight and failed to update some internal documentation, leading me to mark a case as "impossible" when it isn't in the current version of the code. A fix to prevent the crash should be coming along soon.

**Comment by pkgw** [comment]
OK, this should be fixed in the just-released 0.9.0.

**Comment by oovm** [comment]
Thanks, fix confirmed
